App with Elasticsearch doesn't work out-of-the-box with Heroku

##### **Overview of the issue**

I'm updating 21-Points Health as part of the JHipster Mini-Book. When I deploy to Heroku, it fails to configure the Elasticsearch add-on. I experienced a [similar issue](https://github.com/jhipster/generator-jhipster/issues/7810) when I updated the book for JHipster v5.

When I run `jhipster heroku`, it does mention something about Elasticsearch in that "no new Elasticsearch addon created".

```
Provisioning addons

Provisioning database addon heroku-postgresql --as DATABASE

No suitable cache addon for cacheprovider ehcache available.

Creating Heroku deployment files
    force .yo-rc-global.json
   create Procfile
    force .yo-rc.json
   create system.properties
   create gradle/heroku.gradle
 conflict build.gradle
? Overwrite build.gradle? (ynarxdeiH) No new Elasticsearch addon created
? Overwrite build.gradle? overwrite this and all others
```

When I tail the logs after startup, it basically says the [Bonsai Elasticsearch add-on](https://elements.heroku.com/addons/bonsai) wasn't configured.

```
2022-05-13T15:31:15.077985+00:00 app[web.1]: Caused by: java.net.UnknownHostException: 
  ${BONSAI_URL}: Name or service not known
```

I tried installing the add-on manually:

```
$ heroku addons:create bonsai:sandbox-6
Creating bonsai:sandbox-6 on ⬢ health-by-points... free
Your Bonsai cluster is being provisioned.

Please run `heroku addons:open bonsai` and navigate to the Notifications tab to enter contact information for this resource.

Please contact support@bonsai.io for help.
bonsai-clean-03403 is being created in the background. The app will restart when complete...
Use heroku addons:info bonsai-clean-03403 to check creation progress
Use heroku addons:docs bonsai to view documentation
```

Then, I ran `heroku restart`. Now it fails with:

```
... app[web.1]: Caused by: org.elasticsearch.ElasticsearchException: Invalid or missing build flavor [oss]
... app[web.1]: at org.elasticsearch.client.RestHighLevelClient.performClientRequest(RestHighLevelClient.java:2084)
... app[web.1]: at org.elasticsearch.client.RestHighLevelClient.internalPerformRequest(RestHighLevelClient.java:1732)
... app[web.1]: at org.elasticsearch.client.RestHighLevelClient.performRequest(RestHighLevelClient.java:1717)
... app[web.1]: at org.elasticsearch.client.IndicesClient.exists(IndicesClient.java:963)
... app[web.1]: at org.springframework.data.elasticsearch.core.RestIndexTemplate.lambda$doExists$2(RestIndexTemplate.java:103)
... app[web.1]: at org.springframework.data.elasticsearch.core.ElasticsearchRestTemplate.execute(ElasticsearchRestTemplate.java:408)
```

If I google the error, I land at https://github.com/elastic/elasticsearch/issues/76091. 

##### **Motivation for or Use Case**

Elasticsearch + Heroku should work out-of-the-box. 

##### **Reproduce the error**

Create a new app with Elasticsearch and deploy it to Heroku with `jhipster heroku`. 

##### **Related issues**

We experienced a similar issue in June 2018: https://github.com/jhipster/generator-jhipster/issues/7810

##### **Suggest a Fix**

Fix the Heroku sub-generator so it works.

##### **JHipster Version(s)**

```
twenty-one-points@0.0.1-SNAPSHOT /Users/mraible/dev/21-points
└── generator-jhipster@7.8.1

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "authenticationType": "jwt",
    "baseName": "TwentyOnePoints",
    "blueprints": [],
    "buildTool": "gradle",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "clientPackageManager": "npm",
    "clientTheme": "none",
    "clientThemeVariant": "",
    "creationTimestamp": 1651460089159,
    "cypressCoverage": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "devServerPort": 4200,
    "dtoSuffix": "DTO",
    "enableGradleEnterprise": false,
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "entitySuffix": "",
    "herokuAppName": "health-by-pints",
    "herokuDeployType": "git",
    "herokuJavaVersion": "11",
    "jhiPrefix": "jhi",
    "jhipsterVersion": "7.8.1",
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",
    "languages": ["en", "fr"],
    "messageBroker": false,
    "nativeLanguage": "en",
    "otherModules": [],
    "packageName": "org.jhipster.health",
    "pages": [],
    "prodDatabaseType": "postgresql",
    "reactive": false,
    "searchEngine": "elasticsearch",
    "serverPort": "8080",
    "serverSideOptions": ["searchEngine:elasticsearch"],
    "serviceDiscoveryType": "no",
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "skipUserManagement": false,
    "testFrameworks": ["cypress"],
    "websocket": false,
    "withAdminUi": true
  }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>


##### **Environment and Tools**

openjdk version "17.0.3" 2022-04-19
OpenJDK Runtime Environment GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06)
OpenJDK 64-Bit Server VM GraalVM CE 22.1.0 (build 17.0.3+7-jvmci-22.1-b06, mixed mode, sharing)

git version 2.32.0 (Apple Git-132)

node: v16.15.0

npm: 8.6.0

Docker version 20.10.14, build a224086

Docker Compose version v2.5.0

##### **Browsers and Operating System**

Chrome and macOS Monterey

- [x] Checking this box is mandatory (this is just to show you read everything)

--- a/generators/heroku/index.js
+++ b/generators/heroku/index.js
@@ -489,13 +489,15 @@ module.exports = class extends BaseBlueprintGenerator {
 
         this.log(chalk.bold('\nProvisioning addons'));
         if (this.searchEngine === ELASTICSEARCH) {
+          this.log(chalk.bold('\nProvisioning bonsai elasticsearch addon'));
           ChildProcess.exec(
-            `heroku addons:create bonsai:sandbox --as BONSAI --app ${this.herokuAppName}`,
-            addonCreateCallback.bind(this, 'Elasticsearch')
-          );
+            `heroku addons:create bonsai:sandbox-6 --as BONSAI --app ${this.herokuAppName}`, (err, stdout, stderr) => {
+              addonCreateCallback.bind('Elasticsearch', err, stdout, stderr);
+            });
         }
 
         if (this.useOkta) {
+          this.log(chalk.bold('\nProvisioning okta addon'));
           ChildProcess.exec(`heroku addons:create okta --app ${this.herokuAppName}`, (err, stdout, stderr) => {
             addonCreateCallback('Okta', err, stdout, stderr);
           });

--- a/generators/heroku/index.js
+++ b/generators/heroku/index.js
@@ -490,10 +490,9 @@ module.exports = class extends BaseBlueprintGenerator {
         this.log(chalk.bold('\nProvisioning addons'));
         if (this.searchEngine === ELASTICSEARCH) {
           this.log(chalk.bold('\nProvisioning bonsai elasticsearch addon'));
-          ChildProcess.exec(
-            `heroku addons:create bonsai:sandbox-6 --as BONSAI --app ${this.herokuAppName}`, (err, stdout, stderr) => {
-              addonCreateCallback.bind('Elasticsearch', err, stdout, stderr);
-            });
+          ChildProcess.exec(`heroku addons:create bonsai:sandbox-6 --as BONSAI --app ${this.herokuAppName}`, (err, stdout, stderr) => {
+            addonCreateCallback.bind('Elasticsearch', err, stdout, stderr);
+          });
         }
 
         if (this.useOkta) {

--- a/generators/heroku/templates/application-heroku.yml.ejs
+++ b/generators/heroku/templates/application-heroku.yml.ejs
@@ -76,8 +76,7 @@ spring:
 <%_ } _%>
 <%_ if (searchEngineElasticsearch) { _%>
   elasticsearch:
-    rest:
-      uris: ${BONSAI_URL}
+    uris: ${BONSAI_URL}
 <%_ } _%>
 <%_ if (useOkta) { _%>
   security:

--- a/generators/heroku/templates/heroku.gradle.ejs
+++ b/generators/heroku/templates/heroku.gradle.ejs
@@ -22,7 +22,25 @@ heroku {
     appName = "<%= herokuAppName %>"
     buildpacks = ["heroku/jvm"]
 }
+
+
+
 <%_ if (herokuDeployType === 'git') { _%>
+
+if (System.getenv("DYNO") != null) {
+    configurations {
+        all {
+            resolutionStrategy {
+                // force dependency version as used bonsai add-on as of now only supports 7.10.x
+                // https://github.com/jhipster/generator-jhipster/issues/18650
+                force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
+                force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
+                force 'org.elasticsearch:elasticsearch:7.10.2'
+                force 'org.elasticsearch:plugin:transport-netty4-client:7.10.2'
+            }
+        }
+    }
+}
 
 // Task stage is used by Heroku, see also
 // https://devcenter.heroku.com/articles/deploying-gradle-apps-on-heroku
@@ -38,4 +56,17 @@ gradle.taskGraph.whenReady {taskGraph ->
         }
     }
 }
-<%_ } _%>
+<%_ } else { _%>
+configurations {
+    all {
+        resolutionStrategy {
+            // force dependency version as used bonsai add-on as of now only supports 7.10.x
+            // https://github.com/jhipster/generator-jhipster/issues/18650
+            force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
+            force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
+            force 'org.elasticsearch:elasticsearch:7.10.2'
+            force 'org.elasticsearch:plugin:transport-netty4-client:7.10.2'
+        }
+    }
+}
+<%_ }  _%>

--- a/generators/heroku/templates/heroku.gradle.ejs
+++ b/generators/heroku/templates/heroku.gradle.ejs
@@ -23,20 +23,18 @@ heroku {
     buildpacks = ["heroku/jvm"]
 }
 
-
-
+// force dependency version as used bonsai add-on as of now only supports 7.10.x
+// https://github.com/jhipster/generator-jhipster/issues/18650
+def bonsaiElasticSearchVersion = "7.10.2"
 <%_ if (herokuDeployType === 'git') { _%>
-
 if (System.getenv("DYNO") != null) {
     configurations {
         all {
             resolutionStrategy {
-                // force dependency version as used bonsai add-on as of now only supports 7.10.x
-                // https://github.com/jhipster/generator-jhipster/issues/18650
-                force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
-                force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
-                force 'org.elasticsearch:elasticsearch:7.10.2'
-                force 'org.elasticsearch.plugin:transport-netty4-client:7.10.2'
+                force "org.elasticsearch.client:elasticsearch-rest-client:${bonsaiElasticSearchVersion}"
+                force "org.elasticsearch.client:elasticsearch-rest-high-level-client:${bonsaiElasticSearchVersion}"
+                force "org.elasticsearch:elasticsearch:${bonsaiElasticSearchVersion}"
+                force "org.elasticsearch.plugin:transport-netty4-client:${bonsaiElasticSearchVersion}"
             }
         }
     }
@@ -60,12 +58,10 @@ gradle.taskGraph.whenReady {taskGraph ->
 configurations {
     all {
         resolutionStrategy {
-            // force dependency version as used bonsai add-on as of now only supports 7.10.x
-            // https://github.com/jhipster/generator-jhipster/issues/18650
-            force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
-            force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
-            force 'org.elasticsearch:elasticsearch:7.10.2'
-            force 'org.elasticsearch.plugin:transport-netty4-client:7.10.2'
+            force "org.elasticsearch.client:elasticsearch-rest-client:${bonsaiElasticSearchVersion}"
+            force "org.elasticsearch.client:elasticsearch-rest-high-level-client:${bonsaiElasticSearchVersion}"
+            force "org.elasticsearch:elasticsearch:${bonsaiElasticSearchVersion}"
+            force "org.elasticsearch.plugin:transport-netty4-client:${bonsaiElasticSearchVersion}"
         }
     }
 }

--- a/generators/heroku/templates/heroku.gradle.ejs
+++ b/generators/heroku/templates/heroku.gradle.ejs
@@ -36,7 +36,7 @@ if (System.getenv("DYNO") != null) {
                 force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
                 force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
                 force 'org.elasticsearch:elasticsearch:7.10.2'
-                force 'org.elasticsearch:plugin:transport-netty4-client:7.10.2'
+                force 'org.elasticsearch.plugin:transport-netty4-client:7.10.2'
             }
         }
     }
@@ -65,7 +65,7 @@ configurations {
             force 'org.elasticsearch.client:elasticsearch-rest-client:7.10.2'
             force 'org.elasticsearch.client:elasticsearch-rest-high-level-client:7.10.2'
             force 'org.elasticsearch:elasticsearch:7.10.2'
-            force 'org.elasticsearch:plugin:transport-netty4-client:7.10.2'
+            force 'org.elasticsearch.plugin:transport-netty4-client:7.10.2'
         }
     }
 }

--- a/generators/heroku/templates/pom-profile.xml.ejs
+++ b/generators/heroku/templates/pom-profile.xml.ejs
@@ -1,24 +1,27 @@
+            <properties>
+                <bonsai.elasticsearch.version>7.10.2</bonsai.elasticsearch.version>
+            </properties>
             <dependencyManagement>
                 <dependencies>
                     <dependency>
                         <groupId>org.elasticsearch.client</groupId>
                         <artifactId>elasticsearch-rest-client</artifactId>
-                        <version>7.10.2</version>
+                        <version>${bonsai.elasticsearch.version}</version>
                     </dependency>
                     <dependency>
                         <groupId>org.elasticsearch.client</groupId>
                         <artifactId>elasticsearch-rest-high-level-client</artifactId>
-                        <version>7.10.2</version>
+                        <version>${bonsai.elasticsearch.version}</version>
                     </dependency>
                     <dependency>
                         <groupId>org.elasticsearch</groupId>
                         <artifactId>elasticsearch</artifactId>
-                        <version>7.10.2</version>
+                        <version>${bonsai.elasticsearch.version}</version>
                     </dependency>
                     <dependency>
                         <groupId>org.elasticsearch.plugin</groupId>
                         <artifactId>transport-netty4-client</artifactId>
-                        <version>7.10.2</version>
+                        <version>${bonsai.elasticsearch.version}</version>
                     </dependency>
                 </dependencies>
             </dependencyManagement>

--- a/generators/heroku/templates/pom-profile.xml.ejs
+++ b/generators/heroku/templates/pom-profile.xml.ejs
@@ -1,3 +1,27 @@
+            <dependencyManagement>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.elasticsearch.client</groupId>
+                        <artifactId>elasticsearch-rest-client</artifactId>
+                        <version>7.10.2</version>
+                    </dependency>
+                    <dependency>
+                        <groupId>org.elasticsearch.client</groupId>
+                        <artifactId>elasticsearch-rest-high-level-client</artifactId>
+                        <version>7.10.2</version>
+                    </dependency>
+                    <dependency>
+                        <groupId>org.elasticsearch</groupId>
+                        <artifactId>elasticsearch</artifactId>
+                        <version>7.10.2</version>
+                    </dependency>
+                    <dependency>
+                        <groupId>org.elasticsearch.plugin</groupId>
+                        <artifactId>transport-netty4-client</artifactId>
+                        <version>7.10.2</version>
+                    </dependency>
+                </dependencies>
+            </dependencyManagement>
             <build>
                 <plugins>
                     <plugin>

--- a/generators/heroku/templates/pom-profile.xml.ejs
+++ b/generators/heroku/templates/pom-profile.xml.ejs
@@ -1,3 +1,5 @@
+            <!-- force dependency version as used bonsai add-on as of now only supports 7.10.x -->
+            <!-- https://github.com/jhipster/generator-jhipster/issues/18650 -->
             <properties>
                 <bonsai.elasticsearch.version>7.10.2</bonsai.elasticsearch.version>
             </properties>

